query/functions.go

```diff
@@ -566,6 +566,76 @@ func (r *UnsignedMovingAverageReducer) Emit() []FloatPoint {
 	}
 }
 
+// ExponentialMovingAverageReducer calculates the
+// exponential moving average of the aggregated points.
+// It can aggregate both float and integer points.
+type ExponentialMovingAverageReducer struct {
+	decay float64 // the decay factor ( == 1/(age + 1) )
+	value float64 // current value of the moving average
+	count uint32  // number of aggregated samples
+	time  int64   // time of last sample
+}
+
+// NewExponentialMovingAverageReducer creates a new ExponentialMovingAverageReducer.
+func NewExponentialMovingAverageReducer(age int) *ExponentialMovingAverageReducer {
+	return &ExponentialMovingAverageReducer{
+		decay: 2 / (float64(age) + 1),
+	}
+}
+
+// exponentialMovingAverageWarmupSamples is the number of samples used to warm up
+// the exponential moving average calculation.
+//
+// For best results, the moving average should not be initialized to the
+// samples it sees immediately.  The book "Production and Operations
+// Analysis" by Steven Nahmias suggests initializing the moving average to
+// the mean of the first 10 samples.  Until the exponential moving average
+// reducer has seen this many samples, it is not "ready" to be queried for the
+// value of the moving average.
+const exponentialMovingAverageWarmupSamples = 10
```

Comment:
This is a very subjective value. For example a recommendation I see all the time, and what I personally use, is to use the period count as the warmup size. Meaning if your taking the EMA over a 20-period span, your warmup is 20.

Reply:
True. I know it's bad to use magic numbers and perhaps it'd better be configurable by the user, but this value is usually very insignificant when looking at longer time spans. I think it just prevents having a EMA graph that starts with some crazy outliers and then smooths.

Reply:
Like I said, I looked for similar implementations and took the one that at least mentions some sources on their magic numbers.

Reply:
I would argue the opposite, that the value is very significant. If you're taking an EMA with a period in the range of thousands or more, 10 periods is insignificant. It might as well be 0. The first chunk of your results aren't going to be smooth at all.

Reply:
@phemmer  EMA doesn't really have a "period". All samples incorporate all past ones; there is just a decay factor. So let's say a sample that goes like `[1000, 0, 1000, 0, 1000, 0, ...]`, will have a very ugly first 5-10 samples, whereas with a small warmup period, it startss with 500 and will go up and down slightly when it keeps iterating.

Or f.e. a sample that is quite regular but has an outlier like every 100 samples. If you're unlucky and this outlier is in your first few samples (`[3, 6, 100, 2, 2, 1, 4, ...]`). Without the warmup period, that will look ugly. With the warmup period, the 100 will still put the EMA too high in the beginning, but at least it won't jump from ~5 to ~100 and start falling by then.

I'm no expert in the maths behind this, don't get me wrong. I just understood that the value is not that important because EMA's are mainly used for longer series; but that it's just a convenience to prevent ugly data.

If you have a concrete proposal (like making it 0) that would make this change more acceptable to the maintainers, I'd be willing to change the implementation.